File: AlgorithmStudy/11주차/BOJ7785/김재환7785.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.HashSet;
import java.util.ArrayList;
import java.util.Collections;

public class Main {
    public static void main(String[] args) throws Exception {
        // 입출력 최적화를 위한 BufferedReader
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // 첫 줄: 기록 개수
        int n = Integer.parseInt(br.readLine());

        // 현재 회사에 있는 사람들의 이름을 저장할 Set
        // Set을 쓰면 enter 시 중복 없이 저장, leave 시 빠르게 삭제 가능
        HashSet<String> set = new HashSet<>();

        // n개의 로그 처리
        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());

            // 이름
            String name = st.nextToken();

            // 행동: "enter" 또는 "leave"
            String action = st.nextToken();

            // "enter"이면 Set에 저장
            if (action.equals("enter")) {
                set.add(name);
            } else {
                // "leave"이면 Set에서 제거
                // remove는 존재 여부 체크 없이 호출해도 안전
                set.remove(name);
            }
        }

        // 현재 Set 안에 남아있는 사람 이름들을 List로 옮긴다
        ArrayList<String> list = new ArrayList<>(set);

        // 사전 순 역순 정렬
        // reverseOrder()를 사용하면 내림차순 정렬이 된다
        Collections.sort(list, Collections.reverseOrder());

        // 정렬된 이름을 한 줄씩 출력
        StringBuilder sb = new StringBuilder();
        for (String name : list) {
            sb.append(name).append('\n');
        }

        System.out.print(sb);
    }
}

File: AlgorithmStudy/11주차/BOJ7785/김재환7785.py
import sys

# 많은 입력을 빠르게 처리하기 위해
input = sys.stdin.readline

# 로그 개수
n = int(input())

# 현재 회사에 있는 사람을 저장할 집합
# enter -> 추가, leave -> 삭제
now = set()

for _ in range(n):
    name, action = input().split()

    # enter이면 set에 추가
    if action == "enter":
        now.add(name)
    else:
        # leave이면 set에서 삭제
        now.remove(name)

# 사전 순의 역순으로 정렬
# sorted()에 reverse=True를 붙이면 내림차순 정렬
ans = sorted(now, reverse=True)

# 출력
for name in ans:
    print(name)
